File: labRepositoryJava/src/main/java/org/example/repository/ConcertDBRepository.java
package org.example.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.model.Concert;
import org.example.repository.exceptions.RepositoryException;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Properties;

public class ConcertDBRepository implements ConcertRepository{

    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();

    public ConcertDBRepository(Properties properties){
        logger.info("Initializing ConcertDBRepository with properties: {} ",properties);
        dbUtils=new JdbcUtils(properties);
    }

    @Override
    public Concert findOne(Integer integer) {
        logger.traceEntry("find concert by id task{}");
        Connection con=dbUtils.getConnection();
        Concert concert=null;
        try(PreparedStatement preparedStatement=con.prepareStatement("select * from concerts where id=?")){
            preparedStatement.setInt(1,integer);
            try(ResultSet resultSet=preparedStatement.executeQuery()){
                while(resultSet.next()){
                    concert=new Concert(resultSet.getInt("id"),resultSet.getString("concertDate"),resultSet.getString("concertLocation"), resultSet.getInt("ticketsAvailable"),resultSet.getInt("ticketsSold"),resultSet.getInt("artistID"),resultSet.getString("startingTime"));
                }
            }
        }
        catch(SQLException e){
            logger.error(e);
            System.err.println("ERROR DB"+ e);
        }
        logger.traceExit(concert);
        if(concert ==null)
            throw new RepositoryException("no concert has been found with the given id!");
        return concert;
    }

    @Override
    public Iterable<Concert> findAll() {
        logger.traceEntry("get all concerts task{}");
        Connection con= dbUtils.getConnection();
        List<Concert> concerts=new ArrayList<>();
        try(PreparedStatement preparedStatement=con.prepareStatement("select * from concerts")){
            try(ResultSet resultSet=preparedStatement.executeQuery()){
                while(resultSet.next()){
                    int id=resultSet.getInt("id");
                    String concertDate=resultSet.getString("concertDate");
                    String concertLocation=resultSet.getString("concertLocation");
                    Integer ticketsAvailable=resultSet.getInt("ticketsAvailable");
                    Integer ticketsSold=resultSet.getInt("ticketsSold");
                    Integer artistId=resultSet.getInt("artistID");
                    String startingTime=resultSet.getString("startingTime");
                    Concert concert=new Concert(id,concertDate,concertLocation,ticketsAvailable,ticketsSold,artistId,startingTime);
                    concerts.add(concert);
                }
            }
        }catch(SQLException e){
            logger.error(e);
            System.err.println("ERROR DB"+ e);
        }
        logger.traceExit(concerts);
        return concerts;

    }

    @Override
    public Concert save(Concert entity) {
        return null;
    }

    @Override
    public Concert delete(Integer integer) {
        return null;
    }

    @Override
    public Concert update(Concert entity) {
        return null;
    }

    @Override
    public Collection<Concert> getConcertsByDate(String date) {
        logger.traceEntry("get concerts by date task {}");
        Connection con= dbUtils.getConnection();
        List<Concert> concerts=new ArrayList<>();
        try(PreparedStatement preparedStatement=con.prepareStatement("select * from concerts where concertDate=?")){
            preparedStatement.setString(1,date);
            try(ResultSet resultSet=preparedStatement.executeQuery()){
                while(resultSet.next()){
                    int id=resultSet.getInt("id");
                    String concertDate=resultSet.getString("concertDate");
                    String concertLocation=resultSet.getString("concertLocation");
                    Integer ticketsAvailable=resultSet.getInt("ticketsAvailable");
                    Integer ticketsSold=resultSet.getInt("ticketsSold");
                    Integer artistId=resultSet.getInt("artistID");
                    String startingTime=resultSet.getString("startingTime");
                    Concert concert=new Concert(id,concertDate,concertLocation,ticketsAvailable,ticketsSold,artistId,startingTime);
                    concerts.add(concert);
                }
            }
        }catch(SQLException e){
            logger.error(e);
            System.err.println("ERROR DB" + e);
        }
        logger.traceExit(concerts);
        if(concerts.size()==0)
            throw new RepositoryException("no concerts have the given date");
        return concerts;
    }

    @Override
    public Integer updateConcertTickets(Integer id, Integer newTicketsAvailable,Integer newTicketsSold) {
        logger.traceEntry("update concert tickets by id {}");
        Connection con=dbUtils.getConnection();
        int result=0;
        try(PreparedStatement preparedStatement=con.prepareStatement("update concerts set ticketsAvailable=?,ticketsSold=? where id=?")){
            preparedStatement.setInt(1,newTicketsAvailable);
            preparedStatement.setInt(2,newTicketsSold);
            preparedStatement.setInt(3,id);
            result=preparedStatement.executeUpdate();
            logger.trace("updated {} objects",result);
        }catch(SQLException e){
            logger.error(e);
            System.err.println("ERROR DB"+ e);
        }
        if(result==0)
            throw new RepositoryException("no concert with the id given");
        return result;
    }

    @Override
    public Concert getConcertByDateLocationStart(String data, String location,String startTime) {
        logger.traceEntry("get concert by date,location and start time");
        Connection con=dbUtils.getConnection();
        Concert concert=null;
        try(PreparedStatement preparedStatement=con.prepareStatement("select * from concerts where concertDate=? and concertLocation=? and startingTime=?")){
            preparedStatement.setString(1,data);
            preparedStatement.setString(2,location);
            preparedStatement.setString(3,startTime);
            try(ResultSet resultSet=preparedStatement.executeQuery()){
                while(resultSet.next()){
                    int id=resultSet.getInt("id");
                    String concertDate=resultSet.getString("concertDate");
                    String concertLocation=resultSet.getString("concertLocation");
                    Integer ticketsAvailable=resultSet.getInt("ticketsAvailable");
                    Integer ticketsSold=resultSet.getInt("ticketsSold");
                    Integer artistId=resultSet.getInt("artistID");
                    String startingTime=resultSet.getString("startingTime");
                    concert=new Concert(id,concertDate,concertLocation,ticketsAvailable,ticketsSold,artistId,startingTime);
                }
            }
        }catch(SQLException e){
            logger.error(e);
            System.out.println("ERROR DB "+e);
        }
        logger.traceExit(concert);
        if(concert==null)
            throw new RepositoryException("no concert with date,location and start time given");
        return concert;
    }
}

File: labSocketsFestivalCSCompleted/labSocketsFestivalCSV2/ConsoleApplication1/persistence/ConcertDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using model;

namespace persistence
{
    public class ConcertDBRepository:IConcertRepository
    {
        
        private IDictionary<string, string> props;

        public ConcertDBRepository(IDictionary<string, string> props)
        {
            this.props = props;
        }
        
        public void Save(Concert entity)
        {
            throw new System.NotImplementedException();
        }

        public Concert FindOne(int id)
        {
            var connection = DbUtils.getConnection(props);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from concerts where id=@id";

                var paramID = command.CreateParameter();
                paramID.ParameterName = "@id";
                paramID.Value = id;
                command.Parameters.Add(paramID);

                using (var dataR = command.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        Concert concert = new Concert(dataR.GetInt32(0), dataR.GetString(1), dataR.GetString(2),
                            dataR.GetInt32(3), dataR.GetInt32(4), dataR.GetInt32(5), dataR.GetString(6));
                        return concert;
                    }
                }
            }
            return null;
        }

        public IEnumerable<Concert> FindAll()
        {
            var connection = DbUtils.getConnection(props);
            IList <Concert> concerts= new List<Concert>();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from concerts";
                using (var dataR = command.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int id = dataR.GetInt16(0);
                        string date = dataR.GetString(1);
                        string concertLocation = dataR.GetString(2);
                        int ticketsAvailable = dataR.GetInt16(3);
                        int ticketsSold = dataR.GetInt16(4);
                        int artistId = dataR.GetInt16(5);
                        string startTime = dataR.GetString(6);
                        Concert concert = new Concert(id, date, concertLocation, ticketsAvailable, ticketsSold, artistId,startTime);
                        concerts.Add(concert);
                    }
                }
            }
            return concerts;
        }

        public void Remove(int id)
        {
            throw new System.NotImplementedException();
        }

        public void Update(int id, Concert e)
        {
            throw new System.NotImplementedException();
        }

        public IEnumerable<Concert> GetConcertsByDate(string date)
        {
            var connection = DbUtils.getConnection(props);
            IList <Concert> concerts= new List<Concert>();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from concerts where date=@date";

                var paramData = command.CreateParameter();
                paramData.ParameterName = "@date";
                paramData.Value = date;
                command.Parameters.Add(paramData);
                
                using (var dataR = command.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int id = dataR.GetInt16(0);
                        string data = dataR.GetString(1);
                        string concertLocation = dataR.GetString(2);
                        int ticketsAvailable = dataR.GetInt16(3);
                        int ticketsSold = dataR.GetInt16(4);
                        int artistId = dataR.GetInt16(5);
                        string startTime = dataR.GetString(6);
                        Concert concert = new Concert(id, date, concertLocation, ticketsAvailable, ticketsSold, artistId,startTime);
                        concerts.Add(concert);
                    }
                }
            }
            return concerts;
        }

        public void UpdateConcertTickets(int id, int newTicketsAvailable, int newTicketsSold)
        {
            var connection = DbUtils.getConnection(props);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "update concerts set ticketsAvailable=@newTicketsAvailable,ticketsSold=@newTicketsSold where id=@id";

                var paramTicketsA = command.CreateParameter();
                paramTicketsA.ParameterName = "@newTicketsAvailable";
                paramTicketsA.Value = newTicketsAvailable;
                command.Parameters.Add(paramTicketsA);

                var paramTicketsS = command.CreateParameter();
                paramTicketsS.ParameterName = "@newTicketsSold";
                paramTicketsS.Value = newTicketsSold;
                command.Parameters.Add(paramTicketsS);
                
                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                {
                    throw new Exception("Not updated!");
                }
            }
        }

        public Concert getConcertByDateLocationStart(string date, string location, string start)
        {
            
            var connection = DbUtils.getConnection(props);
            using (var command = connection.CreateCommand())
            {
                command.CommandText="SELECT * FROM concerts where date=@data and concertLocation=@location and startTime=@start";

                var paramDate = command.CreateParameter();
                paramDate.ParameterName = "@data";
                paramDate.Value = date;
                command.Parameters.Add(paramDate);

                var paramLoc = command.CreateParameter();
                paramLoc.ParameterName = "@location";
                paramLoc.Value = location;
                command.Parameters.Add(paramLoc);

                var paramStart = command.CreateParameter();
                paramStart.ParameterName = "@start";
                paramStart.Value = start;
                command.Parameters.Add(paramStart);

                using (var dataR = command.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        Concert concert = new Concert(dataR.GetInt32(0),dataR.GetString(1),dataR.GetString(2),dataR.GetInt32(3),dataR.GetInt32(4),dataR.GetInt32(5),dataR.GetString(6));
                        return concert;
                    }
                }
            }
            return null;
        }

        public IList<ConcertInfo> getConcertInfos()
        {
            IList<ConcertInfo> concertInfo = new List<ConcertInfo>();
            IDbConnection connection = DbUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText =
                    "select A.name,C.id,C.concertDate,C.concertLocation,C.ticketsAvailable,C.ticketsSold," +
                    "C.startingTime from concerts C inner join artists A on C.artistID=A.id";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        string artist = dataR.GetString(0);
                        int concertId = dataR.GetInt32(1);
                        string date = dataR.GetString(2);
                        string location = dataR.GetString(3);
                        int ticketsAvailable = dataR.GetInt32(4);
                        int ticketsSold = dataR.GetInt32(5);
                        string startTime = dataR.GetString(6);
                        ConcertInfo concertInf = new ConcertInfo(artist, concertId, date, location, ticketsAvailable, ticketsSold, startTime);
                        concertInfo.Add(concertInf);
                    }
                }
            }
            return concertInfo;
        }

        public IList<ConcertInfo> getConcertInfosByDate(string data)
        {
            IList<ConcertInfo> concertInfo = new List<ConcertInfo>();
            IDbConnection connection = DbUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText =
                    "select A.name,C.id,C.concertDate,C.concertLocation,C.ticketsAvailable,C.ticketsSold," +
                    "C.startingTime from concerts C inner join artists A on C.artistID=A.id where C.concertDate=@date";

                var parDate = comm.CreateParameter();
                parDate.ParameterName = "@date";
                parDate.Value = data;
                comm.Parameters.Add(parDate);
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        string artist = dataR.GetString(0);
                        int concertId = dataR.GetInt32(1);
                        string date = dataR.GetString(2);
                        string location = dataR.GetString(3);
                        int ticketsAvailable = dataR.GetInt32(4);
                        int ticketsSold = dataR.GetInt32(5);
                        string startTime = dataR.GetString(6);
                        ConcertInfo concertInf = new ConcertInfo(artist, concertId, date, location, ticketsAvailable, ticketsSold, startTime);
                        concertInfo.Add(concertInf);
                    }
                }
            }
            return concertInfo;
        }
    }
}